File: src/main/scala/robosky/uplands/block/bossroom/ControlBlock.scala
package robosky.uplands.block.bossroom

import net.fabricmc.fabric.api.block.FabricBlockSettings
import net.minecraft.block.{Block, BlockEntityProvider, BlockState, Material}
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.SpawnEggItem
import net.minecraft.sound.BlockSoundGroup
import net.minecraft.state.StateManager
import net.minecraft.state.property.{EnumProperty, Property}
import net.minecraft.util.{ActionResult, Hand}
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.{BlockPos, Direction}
import net.minecraft.world.{BlockView, World}
import robosky.uplands.UplandsMod
import robosky.uplands.item.UplandsItemTags

object ControlBlock {
  val ADJUST: Property[ControlAdjustment] = EnumProperty.of("adjust", classOf[ControlAdjustment])
}

/**
 * Block for the boss controller. The important bits are on the block entity.
 */
class ControlBlock extends Block(
  FabricBlockSettings.of(Material.STONE)
    .sounds(BlockSoundGroup.STONE)
    .strength(-1.0f, 3600000.0f).build()) with BlockEntityProvider {

  override def createBlockEntity(world: BlockView): BlockEntity = new ControlBlockEntity()

  /**
   * Get the control block entity associated with the given position.
   */
  def control(world: World, pos: BlockPos): Option[ControlBlockEntity] =
    Option(world.getBlockEntity(pos)) collect { case ctrl: ControlBlockEntity => ctrl }

  override def appendProperties(builder: StateManager.Builder[Block, BlockState]): Unit = {
    builder.add(ControlBlock.ADJUST)
  }

  private val adjustmentMapping = {
    import net.minecraft.util.math.Direction._
    Array(
      Array(WEST, NORTH, EAST, SOUTH), // DOWN/UP
      Array(WEST, DOWN, EAST, UP), // NORTH/SOUTH
      Array(NORTH, DOWN, SOUTH, UP) // WEST/EAST
    )
  }

  override def onUse(state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: Hand, ctx: BlockHitResult): ActionResult = {
    val shouldEdit = player.isCreativeLevelTwoOp &&
      player.inventory.contains(UplandsItemTags.BOSSROOM_TECHNICAL)
    if (shouldEdit) {
      control(world, pos) foreach {
        ctrl =>
          player.getMainHandStack.getItem match {
            // set boss type
            case egg: SpawnEggItem =>
              ctrl.bossType = egg.getEntityType(null)
            // adjust boss room bounds
            case _ =>
              val hitSide = ctx.getSide
              val hitPos = ctx.getPos
              val u = hitSide match {
                case Direction.WEST | Direction.EAST =>
                  (hitPos.getZ - pos.getZ).abs - 0.5
                case _ =>
                  (hitPos.getX - pos.getX).abs - 0.5
              }
              val v = hitSide match {
                case Direction.DOWN | Direction.UP =>
                  (hitPos.getZ - pos.getZ).abs - 0.5
                case _ =>
                  (hitPos.getY - pos.getY).abs - 0.5
              }
              val rSq = u * u + v * v
              // u and v range from [-0.5, 0.5], this is the inner third
              // of that range, i.e. < 1/6 ^ 2
              val centerThird = 0.17 * 0.17
              if (rSq < centerThird) {
                world.setBlockState(pos, state.cycle(ControlBlock.ADJUST))
              } else {
                val theta = Math.atan2(v, u) / Math.PI
                val dir = adjustmentMapping(hitSide.getId / 2)(((theta + 1.25) * 2).toInt % 4)
                val adj = state.get(ControlBlock.ADJUST)
                val blocks = if (adj == ControlAdjustment.IN) -1 else 1
                ctrl.adjustBounds(dir, blocks)
              }
          }
      }
    }
    if (shouldEdit) ActionResult.SUCCESS else ActionResult.PASS
  }
}

File: src/main/java/robosky/uplands/block/bossroom/ControlBlock.java
package robosky.uplands.block.bossroom;

import robosky.uplands.item.UplandsItemTags;

import net.minecraft.block.Block;
import net.minecraft.block.BlockEntityProvider;
import net.minecraft.block.BlockState;
import net.minecraft.block.Material;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.Item;
import net.minecraft.item.SpawnEggItem;
import net.minecraft.sound.BlockSoundGroup;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.EnumProperty;
import net.minecraft.state.property.Property;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Hand;
import net.minecraft.util.hit.BlockHitResult;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;

import net.fabricmc.fabric.api.block.FabricBlockSettings;

/**
 * Block for the boss controller. The important bits are on the block entity.
 */
public class ControlBlock extends Block implements BlockEntityProvider {

    public static final Property<ControlAdjustment> ADJUST = EnumProperty.of("adjust", ControlAdjustment.class);

    public ControlBlock() {
        super(FabricBlockSettings.of(Material.STONE)
            .sounds(BlockSoundGroup.STONE)
            .strength(-1.0f, 3600000.0f).build());
    }

    @Override
    public BlockEntity createBlockEntity(BlockView world) {
        return new ControlBlockEntity();
    }

    @Override
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
        builder.add(ControlBlock.ADJUST);
    }

    private static final Direction[][] adjustmentMapping = {
        {Direction.WEST, Direction.NORTH, Direction.EAST, Direction.SOUTH}, // DOWN/UP
        {Direction.WEST, Direction.DOWN, Direction.EAST, Direction.UP}, // NORTH/SOUTH
        {Direction.NORTH, Direction.DOWN, Direction.SOUTH, Direction.UP} // WEST/EAST
    };

    @Override
    public ActionResult onUse(BlockState state, World world, BlockPos pos, PlayerEntity player, Hand hand, BlockHitResult ctx) {
        boolean shouldEdit = player.isCreativeLevelTwoOp() &&
            player.inventory.contains(UplandsItemTags.BOSSROOM_TECHNICAL);
        BlockEntity be = world.getBlockEntity(pos);
        if (shouldEdit && be instanceof ControlBlockEntity) {
            ControlBlockEntity ctrl = (ControlBlockEntity) be;
            Item heldItem = player.getMainHandStack().getItem();
            if (heldItem instanceof SpawnEggItem) {
                // set boss type
                ctrl.setBossType(((SpawnEggItem) heldItem).getEntityType(null));
            } else {
                // adjust boss room bounds
                Direction hitSide = ctx.getSide();
                Vec3d hitPos = ctx.getPos();
                double u;
                if (hitSide == Direction.WEST || hitSide == Direction.EAST) {
                    u = Math.abs(hitPos.getZ() - pos.getZ()) - 0.5;
                } else {
                    u = Math.abs(hitPos.getX() - pos.getX()) - 0.5;
                }
                double v;
                if (hitSide == Direction.DOWN || hitSide == Direction.UP) {
                    v = Math.abs(hitPos.getZ() - pos.getZ()) - 0.5;
                } else {
                    v = Math.abs(hitPos.getY() - pos.getY()) - 0.5;
                }
                double rSq = u * u + v * v;
                // u and v range from [-0.5, 0.5], this is the inner third
                // of that range, i.e. < 1/6 ^ 2
                final double centerThird = 0.17 * 0.17;
                if (rSq < centerThird) {
                    world.setBlockState(pos, state.cycle(ControlBlock.ADJUST));
                } else {
                    double theta = Math.atan2(v, u) / Math.PI;
                    Direction dir = adjustmentMapping[hitSide.getId() / 2][((int) ((theta + 1.25) * 2)) % 4];
                    ControlAdjustment adj = state.get(ControlBlock.ADJUST);
                    int blocks = adj == ControlAdjustment.IN ? -1 : 1;
                    ctrl.adjustBounds(dir, blocks);
                }
            }
            return ActionResult.SUCCESS;
        }
        return ActionResult.PASS;
    }
}
